mycodo/utils/method.py

```diff
@@ -1,115 +1,94 @@
-        dt = datetime.timedelta(hours=now.hour,
-                                minutes=now.minute,
-                                seconds=now.second)
-        seconds = dt.total_seconds()
-    else:
-        seconds = second_of_day
-
-    # Shift the entire graph using 0 - 360 to determine the degree
-    if shift_angle:
-        percent_angle = shift_angle/360
-        angle_seconds = percent_angle*seconds_per_day
-        if seconds+angle_seconds > seconds_per_day:
-            seconds_shifted = seconds+angle_seconds-seconds_per_day
-        else:
-            seconds_shifted = seconds+angle_seconds
-        percent_of_day = seconds_shifted/seconds_per_day
-    else:
-        percent_of_day = seconds/seconds_per_day
-
-    x = percent_of_day*(P0[0]-P3[0])
-
-    # First, get the t-value associated with x-value, where t is the
-    # parameterization of the Bezier curve and ranges from 0 to 1.
-    # We need the coefficients of the polynomial describing cubic Bezier
-    # (cubic polynomial in t)
-    coefficients = [-P0[0] + 3*P1[0] - 3*P2[0] + P3[0],
-                    3*P0[0] - 6*P1[0] + 3*P2[0],
-                    -3*P0[0] + 3*P1[0],
-                    P0[0] - x]
-    # Find roots of the polynomial to determine the parameter t
-    roots = np.roots(coefficients)
-    # Find the root which is between 0 and 1, and is also real
-    correct_root = None
-    for root in roots:
-        if np.isreal(root) and 0 <= root <= 1:
-            correct_root = root
-    # Check a valid root was found
-    if correct_root is None:
-        print('Error, no valid root found. Are you sure your Bezier curve '
-              'represents a valid function when projected into the xy-plane?')
-        return 0
-    param_t = correct_root
-    # From the value for the t parameter, find the corresponding y-value
-    # using the formula for cubic Bezier curves
-    y = (1-param_t)**3*P0[1] + 3*(1-param_t)**2*param_t*P1[1] + 3*(1-param_t)*param_t**2*P2[1] + param_t**3*P3[1]
-    assert np.isreal(y)
-    # Typecast y from np.complex128 to float64
-    y = y.real
-    return y
-
 
-def calculate_method_setpoint(method_id, table, this_controller, Method, MethodData, logger):
+    def __init__(self, method_data, method_type, logger=None):
+        """
+        Initializes the method class
+        :param method_data: data queried from method_data table
+        :param method_type: method type from method table
+        :param logger: The logger to use
+        :return: 0 (success) or 1 (error) and a setpoint value
+        """
+        self.logger = logger
+
+        self.method_type = method_type
+
+        self.method_data = method_data
+
+        self.method_data_all = self.method_data.filter(MethodData.output_id.is_(None)).all()
+        self.method_data_first = self.method_data.filter(MethodData.output_id.is_(None)).first()
+        self.method_data_repeat = self.method_data.filter(MethodData.duration_sec == 0).first()
```

Comment:
This initializes the base class of all methods, sharing comming functionality that was originally in the beginning of the single long function.